src/Sylius/Bundle/CoreBundle/DependencyInjection/Compiler/BackwardsCompatibility/ResolveShopUserTargetEntityPass.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\DependencyInjection\Compiler\BackwardsCompatibility;
+
+use Sylius\Component\User\Model\UserInterface;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+
+/**
+ * @internal
+ */
+final class ResolveShopUserTargetEntityPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        try {
+            $resolveTargetEntityListener = $container->findDefinition('doctrine.orm.listeners.resolve_target_entity');
+            $shopUserClass = $container->getParameter('sylius.model.shop_user.class');
+        } catch (InvalidArgumentException $exception) {
+            return;
+        }
+
+        $resolveTargetEntityListener->addMethodCall(
```

Comment:
Are you sure this does not break the AdminUser entity? As far as I see this will replace its target-entity with a ShopUser as well.

Reply:
Before this PR, `Sylius\Component\User\Model\UserInterface` targets `ShopUser` entity (cause it's the last one to define `Sylius\Component\User\Model\UserInterface` as its interface). This will preserve backwards compatibility (could be removed in 2.0).

Reply:
Currently at lunch, but I think I remember the AdminUser mapping using the same interface for its user-relation.

Reply:
That's correct, but shop user comes next and overwrites it :)

Reply:
Isn't `resolveTargetEntity` replacing all occurrences of `OriginalTargetEntity` with `NewTargetEntity` and thus both `AdminUser` and `ShopUser` use the `NewTargetEntity` (ShopUserInterface) as its target-entity?

Reply:
It works like this:

```php
$interfaces = [];
$interfaces[UserInterface::class] = AdminUser::class;
$interfaces[UserInterface::class] = ShopUser::class;

var_dump(str_replace(array_keys($interfaces), array_values($interfaces), UserInterface::class));
// string(8) "ShopUser"
```

There can be only one target entity for a given interface.

Reply:
Wait - I think thats on me. 
I got ResourceBundle config and mappings mixed up. 
The `AdminUser` and `ShopUser` had the same interface configured for its resource, but the `ShopUser` is the only one being used with `UserInterface` in a relation (Customer in this case).

The `AdminUser::class` has no target-entity for nothing, so my worries fizzle.
The `ShopUser::class` is (only) being used by the `Customer` as an inverse relation with target-entity `UserInterface` (which gets corrected anyways) and thus re-mapping `UserInterface` to `ShopUserInterface` (or the currently configured resource-class) in the whole application is totally fine. 

My bad, thanks for your time though :heart: